File: Backtracking/78_Subsets/78_Subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        
        78. Subsets
        Medium
        Given an integer array nums of unique elements, return all possible 
        subsets (the power set).

        The solution set must not contain duplicate subsets. Return the solution in any order.

        Example 1:
        Input: nums = [1,2,3]
        Output: [[],[1],[2],[1,2],[3],[1,3],[2,3],[1,2,3]]
        
        Example 2:
        Input: nums = [0]
        Output: [[],[0]]
        """
        # backtrack() appends a copy of current to output at the beginning of each call. 
        # This way, each possible subset is included in the output list. 
        # Then, it iterates through the remaining elements, includes or excludes each one, and continues recursively.
        
          
        def backtrack(start=0, current=[]):
            # Append a copy of the current subset to the output
            output.append(current[::])
            
            # Iterate through the remaining elements to include/exclude in the current subset
            for i in range(start, len(nums)):
                # Include the current element
                current.append(nums[i])
                
                # Recur for the next elements
                backtrack(i + 1, current)
                
                # Backtrack to exclude the current element
                current.pop()

        output = []
        backtrack()
        return output

File: Backtracking/78_Subsets/78_Subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> current = new ArrayList<>();
        backtrack(0, current, res, nums);
        return res;
    }
    private void backtrack(Integer start, List<Integer> current, List<List<Integer>> res, int[] nums) {
        if(start == nums.length) {
            res.add(new ArrayList<>(current));
            return;
        }
        
        // Include the current element
        current.add(nums[start]);
        backtrack(start+1, current, res, nums);  
        
        // Exclude the current element
        current.remove(current.size()-1);
        backtrack(start+1, current, res, nums);
    }
}
